minhashcuda.cc

```diff
@@ -302,33 +302,47 @@ static std::vector<uint32_t> calc_best_split(
   // We evaluate each variant by the cumulative cost function.
   // Every call to mhcuda_calc() can grow the buffers a little; the cost function
   // optimizes for the number of reallocations first and the imbalance second.
-
+  if (devs.size() == 1) {
+    return {length};
+  }
   uint32_t ideal_split = rows[length] / devs.size();
+  uint32_t previous_row = 0;
   std::vector<std::vector<uint32_t>> variants;
-  for (size_t devi = 0; devi < devs.size(); devi++) {
+  for (size_t devi = 0; devi < devs.size() - 1; devi++) {
```

Comment:
Better:

```
std::upper_bound(rows, rows + length + 1, ideal_split * (devi + 1), rows + length)
// default 4th arg
```

Also would be nice to write a comment that `devi < devs.size() - 1` is because you unconditionally add `length` in the end.

By the way: what happens if `length` is less than the number of devices?

Reply:
got it, gonna change and add the comment. hm yes in that case we would have the same error, since one of the GPUs  would have an empty partition. gonna pin down where that happens and change that.